bundles/org.openhab.binding.pixometer/src/main/java/org/openhab/binding/pixometer/handler/ApiserviceHandler.java

```diff
@@ -0,0 +1,174 @@
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.pixometer.handler;
+
+import static org.openhab.binding.pixometer.internal.PixometerBindingConstants.*;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Properties;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.smarthome.config.core.Configuration;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseBridgeHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.io.net.http.HttpUtil;
+import org.openhab.binding.pixometer.internal.PixometerConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+/**
+ * The {@link ApiserviceHandler} is responsible for handling the api connection and authorization (including token
+ * refresh)
+ *
+ * @author Jerome Luckenbach - Initial contribution
+ */
+public class ApiserviceHandler extends BaseBridgeHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+    private static final int TOKEN_MIN_DIFF_MS = (int) TimeUnit.DAYS.toMillis(2);
+
+    private String authToken;
+    private int refreshInterval;
+    private long tokenExpiryDate;
+
+    public ApiserviceHandler(Bridge bridge) {
+        super(bridge);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // Nothing to handle here currently
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initialize Pixometer Apiservice");
+        updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.NONE, "Waiting to get access-token from api...");
+
+        PixometerConfiguration config = getConfigAs(PixometerConfiguration.class);
+        setRefreshInterval(config.refresh);
+        String user = config.user;
+        String password = config.password;
+        String scope = config.scope;
+
+        // Check expiry date every Day and obtain new access token if difference is less then or equal to 2 days
+        scheduler.scheduleWithFixedDelay(() -> {
+            logger.debug("Checking if new access token is needed...");
+            try {
+                long difference = getTokenExpiryDate() - System.nanoTime();
+                if (difference <= TOKEN_MIN_DIFF_MS) {
+                    obtainAuthTokenAndExpiryDate(user, password, scope);
+                }
+            } catch (RuntimeException r) {
```

Comment:
I most of the time catch RuntimeException in scheduled tasks (Not Exception), because if it will throw this exception the scheduled task will stop. If it was a temporary problem due to bad data it means the user has to restart the binding to get it going again. Code should be written to handle such cases, but it's difficult to thing of everything and this catching would at least let the user know. Normally I also combine this with a status update so the user can directly see something might be wrong. (without the user having to look into the log wondering why nothing updates anymore). In this case that might not directly be needed as the second scheduled task result in the status update in case something fails here, because it will get a error due to not authenticated.

Reply:
Fair point, @Hilbrand.
But what I find offensive here is that the internal code throws REs at many places - I would love to see those being more specific and also then be handled more specifically here.
Ok for me to keep this general catch as a safety net for bugs that aren't yet known.

Reply:
To be honest:
I am not too familiar with the existing Exceptions. I have tried to read through the JavaDoc but i didn't invest that much time yet.

So what would be a well Exception Type for this?
And if i read correct you would keep the RE catch as fallback when a specific Excepütion type wouldn't have been catched?

Reply:
The catching in the scheduled tasks is accepted. All other cases are not needed. That is. The cases were an RuntimeException is thrown it should just update the status at that point. The other catches can be omitted as they will be handled catched either by the scheduled task here or by openHAB itself.

Reply:
Thanks for those explanations.

One last question.
When the `catch` can be omitted for the other cases, should i still use a try finally block and just remove the catch or leave out the try comepletely?
(Sorry for the java rookie questions.)

Reply:
Depends on whether there is any code to put in the finally block. If not, you can clearly remove the whole try block.

Reply:
Ok. I have guessed that already and its in the last commit this way. Thanks.